### PR TITLE
8303820: Simplify type metadata

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch simplifies the TypeMetadata API. TypeMetadata can be used as a side-channel, to attach extra information to a javac type (type annotations, constant values).

While TypeMetadata provides the right knobs to compare types (see `Type::equalsIgnoreMetadata`), which is used uniformly across the type-system related routines (e.g. subtyping, type equality and type containment), the TypeMetadata API is also rather cumbersome to use.

The only way to add a new metadata to a type is to "combine" the metadata into an existing one, which feels odd, and probably too biased by the type annotations requirements.

This patch simplifies this: now TypeMetadata is a simple marker interface. There can be many implementations - one is `Annotations` (used to store type annotations), another is `ConstantValue` (used to store a type's constant value).

A type is then associated with a `List<TypeMetadata>`, and there are methods to add/drop/get metadata. These methods are used to provide support for annotated and constant types.

The resulting code feels simpler: to add a new metadata, one only has to define a new class/record inside TypeMetadata, and that's pretty much it. Javac will truck the metadata around in the correct way (as it did before).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303820](https://bugs.openjdk.org/browse/JDK-8303820): Simplify type metadata

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12924/head:pull/12924` \
`$ git checkout pull/12924`

Update a local copy of the PR: \
`$ git checkout pull/12924` \
`$ git pull https://git.openjdk.org/jdk pull/12924/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12924`

View PR using the GUI difftool: \
`$ git pr show -t 12924`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12924.diff">https://git.openjdk.org/jdk/pull/12924.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java [hunk 1]
```diff
@@ -328,7 +323,7 @@ public <Z> Type map(TypeMapping<Z> mapping) {
      *  and with given constant value
      */
     public Type constType(Object constValue) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:325]: so now any type can have a constant value?

**Comment by vicente-romero-oracle** [comment]
nice patch, nice simplification, just added a question

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java (hunk 1 above)
Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:325]: Yeah - not sure which way to go - in reality we have a also a couple of types which implement this method and just return `this`. Only the toplevel `Type` declaration throws. I'm happy to revert to the old ways if that is preferred - it was mostly a way not to duplicate the method body twice (but that's not a big deal).

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:325]: I would keep the original semantics and we can relax it later if needed, but I don't have a strong preference tbh

**Review by vicente-romero-oracle — APPROVED** [review]
looks good thanks

**Pull request closed without merging**